### PR TITLE
ENH: Support rechunk with some unknown dims

### PR DESCRIPTION
**Pull request by TomAugspurger**
Expands the rechunking methods to support rechunking arrays
with some unknown dimensions, along known dimensions.

Closes https://github.com/dask/dask/issues/2236

### CODE REVIEW

**Comment by TomAugspurger** [comment]
I have a TODO in there for updating the `plan_rechunk` code. The current code happens to already "work" since our block sizes are always NaN, so AFACT, the `block_size_limit` is always using `DEFAULT_BLOCK_SIZE_LIMIT`. If I understand correctly, this isn't optimal, but it's the best we can do with unknown dimensions.

File: dask/array/rechunk.py [hunk 1]
```diff
@@ -129,6 +149,9 @@ def intersect_chunks(old_chunks, new_chunks):
     cmn = cumdims_label(new_chunks, 'n')
     sums = [sum(o) for o in old_chunks]
     sums2 = [sum(n) for n in old_chunks]
+
+    sums = [x for x in sums if x == x]  # filter NaNs
+    sums2 = [x for x in sums2 if x == x]
```
Comment (mrocklin) [inline dask/array/rechunk.py]: Maybe use `math.isnan` just for clarity of intent?

File: dask/array/tests/test_rechunk.py [hunk 2]
```diff
@@ -393,3 +393,79 @@ def test_dont_concatenate_single_chunks(shape, chunks):
...
+    x = dd.from_pandas(pd.DataFrame(np.random.randn(50, 10)), 2).values
+    # result = x.rechunk({1: 5})
+    result = x.rechunk((None, (5, 5)))
+    assert np.isnan(x.chunks[0]).all()
+    assert np.isnan(result.chunks[0]).all()
+    assert result.chunks[1] == (5, 5)
+
+
+def test_rechunk_unknown2():
+    import dask; dask.set_options(get=dask.get)  # noqa
```
Comment (mrocklin) [inline dask/array/tests/test_rechunk.py]: Why was this necessary?  If we need it then we might consider using a context manager instead to ensure cleanup.

Comment (TomAugspurger) [inline dask/array/tests/test_rechunk.py]: Sorry, it was for debugging. I'll remove it.

File: dask/array/rechunk.py [hunk 3]
```diff
@@ -45,7 +45,15 @@ def _breakpoints(cumold, cumnew):
     >>> _breakpoints(new[1], old[1])
     (('n', 0), ('o', 0), ('n', 2), ('n', 4), ('n', 5), ('o', 5))
     """
-    return tuple(sorted(cumold + cumnew, key=itemgetter(1)))
+    merged = cumold + cumnew
+    if any(v != v for _, v in merged):
```
Comment (pitrou) [inline dask/array/rechunk.py]: Rather than complicate all subroutines to support unknown dims, perhaps it's easier to modify the callers?

File: dask/array/rechunk.py [hunk 4]
```diff
@@ -589,7 +619,8 @@ def format_blocks(blocks):
     2*[10] | [5, 6] | 3*[2] | [7]
     """
     assert (isinstance(blocks, tuple) and
-            all(isinstance(x, int) for x in blocks))
+            all(isinstance(x, int) or np.isnan(x)
+                for x in blocks))
```
Comment (mrocklin) [inline dask/array/rechunk.py]: No clue if this is a bottleneck, but `math.isnan` is significantly faster on scalars.

**Comment by TomAugspurger** [comment]
> Rather than complicate all subroutines to support unknown dims, perhaps it's easier to modify the callers?

My latest commit tries to do that. To summarize the changes

1. I extracted some of the logic from `intersect_chunks` into a new helper `_old_to_new`
2. All of the new NaN handling is in the `_old_to_new` function.
3. We call the (unchanged) `_intersect_1d` and `_breakpoints` functions on the known dimensions
4. We enhance the known output with the necessary lists of slices to handle the unknown dimensions.

I'm worried I've missed some edge cases, since I still don't quite understand the `_intersect_1d` and `_breakpoints` methods. But the rechunk tests still pass, so that's something.

File: dask/array/tests/test_rechunk.py [hunk 5]
```diff
@@ -393,3 +393,164 @@ def test_dont_concatenate_single_chunks(shape, chunks):
...
+    x = dd.from_pandas(pd.DataFrame(np.random.randn(50, 10)), 2).values
+    # result = x.rechunk({1: 5})
+    result = x.rechunk((None, (5, 5)))
+    assert np.isnan(x.chunks[0]).all()
+    assert np.isnan(result.chunks[0]).all()
+    assert result.chunks[1] == (5, 5)
+
+
+def test_rechunk_unknown_from_array():
+    import dask; dask.set_options(get=dask.get)  # noqa
```
Comment (mrocklin) [inline dask/array/tests/test_rechunk.py]: This is causing test failures.  You probably want this to be under a context manager.  Or it looks like you might not need this at all.  Typically we use `assert_eq` when we need to compute things, this automatically uses the synchronous scheduler (and checks many other useful things).

File: dask/array/tests/test_rechunk.py [hunk 6]
```diff
@@ -393,3 +393,163 @@ def test_dont_concatenate_single_chunks(shape, chunks):
...
+def test_rechunk_unknown_from_pandas():
+    dd = pytest.importorskip('dask.dataframe')
+    pd = pytest.importorskip('pandas')
+
+    x = dd.from_pandas(pd.DataFrame(np.random.randn(50, 10)), 2).values
+    # result = x.rechunk({1: 5})
+    result = x.rechunk((None, (5, 5)))
+    assert np.isnan(x.chunks[0]).all()
+    assert np.isnan(result.chunks[0]).all()
+    assert result.chunks[1] == (5, 5)
```
Comment (mrocklin) [inline dask/array/tests/test_rechunk.py]: It would be nice to call `assert_eq(x, result)` here if possible.  Same below.

File: dask/array/tests/test_rechunk.py [hunk 7]
```diff
@@ -393,3 +393,163 @@ def test_dont_concatenate_single_chunks(shape, chunks):
...
+        else:
+            assert x == y
+
+
+def test_rechunk_unknown_raises():
+    dd = pytest.importorskip('dask.dataframe')
+
+    x = dd.from_array(da.ones(shape=(10, 10), chunks=(5, 5))).values
+    with pytest.raises(ValueError):
+        x.rechunk((None, (5, 5, 5)))
```
Comment (mrocklin) [inline dask/array/tests/test_rechunk.py]: I think that in other operations like `reshape` we sometimes use `None` as a signal that we don't care what happens to this dimension, but don't want to go through the trouble of specifying it.

Comment (TomAugspurger) [inline dask/array/tests/test_rechunk.py]: Thin this case, the `ValueError` comes from the 1-dimension chunks not adding up (10 -> 15). I do have tests ensuring that `None` works as a dummy for "unchanged", though I see now that I don't have a test like `x.rechunk(((nan, nan), (5, 5)))`, which should work, and `x.rechunk(((nan,), (5, 5)))`, which should fail. I'll add them now.

**Comment by mrocklin** [comment]
Ah my apologies, I didn't catch the extra five in there.

File: dask/array/rechunk.py [hunk 8]
```diff
@@ -430,6 +476,8 @@ def plan_rechunk(old_chunks, new_chunks, itemsize,
         The maximum block size (in bytes) we want to produce during an
         intermediate step
     """
+    # TODO: revisit this for unknown chunks. Is this still correct?
+    # Do we have to worry about the memory sizes not being respected?
```
Comment (pitrou) [inline dask/array/rechunk.py]: You should, unless you don't care about an intermediate result possibly swamping memory and swapping out to disk. This is why I suggested that the current heuristic may not be correct for unknown-sized arrays.

File: dask/array/rechunk.py [hunk 9]
```diff
@@ -105,6 +106,48 @@ def _intersect_1d(breaks):
...
+        raise ValueError('Cannot change dimensions from to %r' % sums2)
+    if not n_missing == n_missing2:
+        raise ValueError('Chunks must be unchanging along unknown dimensions')
+
+    old_to_new = [_intersect_1d(_breakpoints(cm[0], cm[1])) for cm in zip(cmo, cmn)]
+    for idx, missing in enumerate(n_missing):
+        if missing:
+            # Missing dimensions are always unchanged, so old -> new is everything
+            extra = [[(i, slice(0, None))] for i in range(missing)]
+            # TODO: hmm avoid an insert here ideally. Build it up and then append?
```
Comment (pitrou) [inline dask/array/rechunk.py]: `old_to_new` is bound to be short, so list.insert being O(n) shouldn't matter, does it?

**Comment by TomAugspurger** [comment]
d7eaa90 should exclude unknown dimensions from any size calculations. I think that's our desired behavior, since we don't allow rechunking along the unknown dimension.

**Comment by pitrou** [comment]
> d7eaa90 should exclude unknown dimensions from any size calculations. I think that's our desired behavior, since we don't allow rechunking along the unknown dimension.

I don't think so, since intermediate results will include the unknown dimensions and their size therefore depends on the size of those dimensions.

**Comment by TomAugspurger** [comment]
> I don't think so, since intermediate results will include the unknown dimensions and their size therefore depends on the size of those dimensions.

I see... Can we do any better than that while planning the rechunking, if we don't know the size of a chunk along an unknown dimension?

**Comment by pitrou** [comment]
> I see... Can we do any better than that while planning the rechunking, if we don't know the size of a chunk along an unknown dimension?

I don't think so. One radical solution would be to never create intermediates in the rechunking plan, another would be to disable the size limitation.

**Comment by TomAugspurger** [comment]
> I don't think so. One radical solution would be to never create intermediates in the rechunking plan, another would be to disable the size limitation.

@mrocklin do you have a preference between these two? I'm leaning towards not having any intermediates, since what's the point of a block size limit parameter, if you can't actually respect it?

**Comment by mrocklin** [comment]
Not having any intermediates seems fine to me.

**Comment by TomAugspurger** [comment]
@mrocklin all green. I think this is ready.

**Comment by mrocklin** [comment]
@pitrou any further concerns on this?

I will probably not go through the logic changes here personally.

File: dask/array/rechunk.py [hunk 10]
```diff
@@ -429,10 +470,17 @@ def plan_rechunk(old_chunks, new_chunks, itemsize,
...
+    No intermediate steps will be planned if any dimension of ``old_chunks``
+    is unknown.
     """
     ndim = len(new_chunks)
     steps = []
-    if ndim <= 1 or not all(new_chunks):
+    has_nans = [any(math.isnan(y) for y in x) for x in old_chunks]
+
+    if ndim <= 1 or not all(new_chunks) or any(has_nans):
         # Trivial array => no need for an intermediate
```
Comment (pitrou) [inline dask/array/rechunk.py]: Update comment here?

File: dask/array/tests/test_rechunk.py (hunk 6 above)
Comment (pitrou) [inline dask/array/tests/test_rechunk.py]: I see the assert_eq() call is still missing...?

File: dask/array/tests/test_rechunk.py [hunk 11]
```diff
@@ -393,3 +399,174 @@ def test_dont_concatenate_single_chunks(shape, chunks):
...
+    (da.ones(shape=(10, 10), chunks=(10, 10)), (None, 5)),
+    (da.ones(shape=(10, 10), chunks=(10, 10)), {1: 5}),
+    (da.ones(shape=(10, 10), chunks=(10, 10)), (None, (5, 5))),
+
+    (da.ones(shape=(10, 10), chunks=(10, 2)), (None, 5)),
+    (da.ones(shape=(10, 10), chunks=(10, 2)), {1: 5}),
+    (da.ones(shape=(10, 10), chunks=(10, 2)), (None, (5, 5))),
+])
+def test_rechunk_unknown(x, chunks):
+    dd = pytest.importorskip('dask.dataframe')
```
Comment (pitrou) [inline dask/array/tests/test_rechunk.py]: Shouldn't this (and other similar tests) go into the dask.dataframe tests?

**Comment by mrocklin** [comment]
OK to merge this?

**Comment by TomAugspurger** [comment]
Good on my end :)

**Comment by mrocklin** [comment]
@pitrou are you ok with this?
